imap_data_access/file_validation.py

```diff
@@ -0,0 +1,264 @@
+"""Methods for managing and validating filenames and filepaths"""
+from __future__ import annotations
+
+import re
+
+from datetime import datetime
+from pathlib import Path
+
+VALID_INSTRUMENTS = {
+    "codice",
+    "glows",
+    "hit",
+    "hi-45",
+    "hi-90",
+    "idex",
+    "lo",
+    "mag",
+    "swapi",
+    "swe",
+    "ultra-45",
+    "ultra-90",
+}
+
+VALID_DATALEVELS = {"l0", "l1", "l1a", "l1b", "l1c", "l1d", "l2"}
+
+VALID_FILE_EXTENSION = {"pkts", "cdf"}
+
+FILENAME_CONVENTION = (
+    "<mission>_<instrument>_<datalevel>_<descriptor>_"
+    "<startdate>_<enddate>_<version>.<extension>"
+)
+
+
+def extract_filename_components(filename: str | Path):
+    """
+    Extracts all components from filename. Does not validate instrument or level.
+
+    Will return a dictionary with the following keys:
+    { instrument, datalevel, descriptor, startdate, enddate, version, extension }
+
+    If a match is not found, a ValueError will be raised.
+
+    Parameters
+    ----------
+    filename : Path or str
+        Path of dependency data.
+
+    Returns
+    -------
+    components : dict
+        Dictionary containing components.
+
+    """
+    pattern = (
+        r"^imap_"
+        r"(?P<instrument>[^_]+)_"
+        r"(?P<datalevel>[^_]+)_"
+        r"(?P<descriptor>[^_]+)_"
+        r"(?P<startdate>\d{8})_"
+        r"(?P<enddate>\d{8})_"
+        r"(?P<version>v\d{2}-\d{2})"
+        r"\.(cdf|pkts)$"
+    )
+    if isinstance(filename, Path):
+        filename = filename.name
+
+    match = re.match(pattern, filename)
+    if match is None:
+        raise ValueError(f"Filename {filename} does not match expected pattern")
+
+    components = match.groupdict()
+    components["extension"] = match.group(7)
+    components["mission"] = "imap"
+    return components
+
+
+def construct_upload_path(filename: str | Path) -> Path:
+    """
+    Given the filename, construct the expected upload path.
+
+    Example path: imap/mag/l1a/2021/01/imap_mag_l1a_burst_20210101_20210102_v01-01.cdf
+
+    Parameters
+    ----------
+    filename: str | Path
+        File name or path to create the path from. Should match IMAP file convention:
+        <mission>_<instrument>_<datalevel>_<descriptor>_<startdate>_<enddate>_<version>.<extension>
+
+    Returns
+    -------
+    Path
+        File path generated from the filename.
+
+    """
+
+    components = extract_filename_components(filename)
+
+    return (
+        Path(
+            f"{components['mission']}/{components['instrument']}/"
+            f"{components['datalevel']}/{components['startdate'][:4]}/"
+            f"{components['startdate'][4:6]}"
+        )
+        / filename
+    )
+
+
+class InvalidScienceFileError(Exception):
```

Comment:
I'm a bit confused from here on down. I think these are two different ways of approaching the same problem, so are we choosing one or the other, or using both for some reason?

Reply:
The main issue is that this is combining two ways of solving this from sds-data-manager. The functions are from a different file than the class. I was having a lot of trouble deciding which way would be better, if functions should be static inside the class or stand alone, etc. What are people's thoughts on a class vs standalone functions? This code would be used in different ways, so sometimes the class is better and sometimes the standalone methods are better.